### ISSUE TITLE
some questions about ssd in mamba2

### ISSUE THREAD

**Issue by shawnnjupt** [state-spaces/mamba#709]
![Image](https://github.com/user-attachments/assets/8c8a5071-232a-4dbd-8980-9a9b940ebdec)

figure 1：mamba2 ssd 

reference code [[mamba/mamba_ssm/modules/ssd_minimal.py at main · state-spaces/mamba](https://github.com/state-spaces/mamba/blob/main/mamba_ssm/modules/ssd_minimal.py)](https://github.com/state-spaces/mamba/blob/main/mamba_ssm/modules/ssd_minimal.py)

I've made a slight change to the test_correctness() function to match the mamba2-130M settings

```python
# Copyright (c) 2024, Albert Gu and Tri Dao.
"""Minimal implementation of SSD.

This is the same as Listing 1 from the paper.
"""
# Copyright (c) 2024, Albert Gu and Tri Dao.
"""Minimal implementation of SSD.

This is the same as Listing 1 from the paper.
"""

import torch
import torch.nn.functional as F
from einops import rearrange, repeat

from mamba_ssm.ops.triton.ssd_combined import mamba_chunk_scan_combined


def segsum_unstable(x):
    """Naive segment sum calculation."""
    T = x.size(-1)
    x_cumsum = torch.cumsum(x, dim=-1)
    x_segsum = x_cumsum[..., :, None] - x_cumsum[..., None, :]
    mask = torch.tril(torch.ones(T, T, device=x.device, dtype=bool), diagonal=0)
    x_segsum = x_segsum.masked_fill(~mask, -torch.inf)
    return x_segsum

def segsum(x):
    """More stable segment sum calculation."""
    T = x.size(-1)
    x = repeat(x, "... d -> ... d e", e=T)
    mask = torch.tril(torch.ones(T, T, device=x.device, dtype=bool), diagonal=-1)
    x = x.masked_fill(~mask, 0)
    x_segsum = torch.cumsum(x, dim=-2)
    mask = torch.tril(torch.ones(T, T, device=x.device, dtype=bool), diagonal=0)
    x_segsum = x_segsum.masked_fill(~mask, -torch.inf)
    return x_segsum

def ssd_minimal_discrete(X, A, B, C, block_len, initial_states=None):
    """
    Arguments:
        X: (batch, length, n_heads, d_head)
        A: (batch, length, n_heads)
        B: (batch, length, n_heads, d_state)
        C: (batch, length, n_heads, d_state)
    Return:
        Y: (batch, length, n_heads, d_head)
    """
    assert X.dtype == A.dtype == B.dtype == C.dtype
    assert X.shape[1] % block_len == 0

    # Rearrange into blocks/chunks
    X, A, B, C = [rearrange(x, "b (c l) ... -> b c l ...", l=block_len) for x in (X, A, B, C)]

    A = rearrange(A, "b c l h -> b h c l")
#Marking Point 1！
    A_cumsum = torch.cumsum(A, dim=-1)
#Marking Point 1！
    # 1. Compute the output for each intra-chunk (diagonal blocks)
    L = torch.exp(segsum(A))
#Marking Point 2！
    Y_diag  = torch.einsum("bclhn,bcshn,bhcls,bcshp->bclhp", C, B, L, X)
#Marking Point 2！
    # 2. Compute the state for each intra-chunk
    # (right term of low-rank factorization of off-diagonal blocks; B terms)
#Marking Point 3！
    decay_states = torch.exp((A_cumsum[:, :, :, -1:] - A_cumsum))
#Marking Point 3！
    states = torch.einsum("bclhn,bhcl,bclhp->bchpn", B, decay_states, X)

    # 3. Compute the inter-chunk SSM recurrence; produces correct SSM states at chunk boundaries
    # (middle term of factorization of off-diag blocks; A terms)
#Marking Point 4！
    if initial_states is None:
        initial_states = torch.zeros_like(states[:, :1])
    states = torch.cat([initial_states, states], dim=1)
#Marking Point 4！
#Marking Point 5！
    decay_chunk = torch.exp(segsum(F.pad(A_cumsum[:, :, :, -1], (1, 0))))
#Marking Point 5！
    new_states = torch.einsum("bhzc,bchpn->bzhpn", decay_chunk, states)
    states, final_state = new_states[:, :-1], new_states[:, -1]

    # 4. Compute state -> output conversion per chunk
    # (left term of low-rank factorization of off-diagonal blocks; C terms)
    state_decay_out = torch.exp(A_cumsum)
    Y_off = torch.einsum('bclhn,bchpn,bhcl->bclhp', C, states, state_decay_out)

    # Add output of intra-chunk and inter-chunk terms (diagonal and off-diagonal blocks)
    Y = rearrange(Y_diag+Y_off, "b c l h p -> b (c l) h p")
    return Y, final_state


# Simple test
def test_correctness():
    torch.manual_seed(42)

    ## Dimensions
    # Denoted (B, T, Q, D, P) in the paper
    batch, seqlen, chunk_size, dim, headdim = 1, 256, 64, 1536, 64
    nheads = dim // headdim  # (H) in the paper
    ngroups = 1 # (G) in the paper
    dstate = 128  # (N) in the paper
    dtype = torch.float32
    device = "cuda"

    x = torch.randn(batch, seqlen, nheads, headdim, dtype=dtype, device=device)
    dt = F.softplus(torch.randn(batch, seqlen, nheads, dtype=torch.float32, device=device) - 4).requires_grad_()
    A = (-torch.exp(torch.rand(nheads, dtype=torch.float32, device=device))).requires_grad_()
    B = torch.randn(batch, seqlen, ngroups, dstate, dtype=dtype, device=device)
    C = torch.randn(batch, seqlen, ngroups, dstate, dtype=dtype, device=device)
    D = torch.randn(nheads, dtype=dtype, device=device)

    # Comparing fused version and minimal version
    y = mamba_chunk_scan_combined(x, dt, A, B, C, chunk_size, D=None)
    y_min, _ = ssd_minimal_discrete(x*dt.unsqueeze(-1), A*dt, B, C, chunk_size)
```

Modified: Changed seqlen to 256 (chunk_size=64, here it is changed to 4 times chunk_size)，change  dim to 1536, so that it is divided into 24 heads, and d_state is changed to 128

I printed out the torch.size of each internal calculation and analyzed it, and then I have the following questions:

## 1、In the mark 1 in the code, I found that the size of A is [1,24,4,64], I don't know if my understanding of the following is correct

According to my understanding, A here corresponds to Figure 1 (the top of this article), that is, each small grid corresponds to the parameters in the 64 rows of the last dimension of the A matrix , and then there are such 4x64 row parameters, which can be marked as a1... a63 /a64... a127/a128 ... a191/a192... a255. And then there are  24 heads.

If this is correct, then for the Y_diag of marker 2, the size of this is (Y_diag_size= torch. Size([1, 4, 64, 24, 64])). 
Also according to Figure 1, in 256 seqlen ,we will find four diagonal cells multiplied by x, and each is a matrix of 64x64, and then there are 24 heads.

## 2, mark 3 is finding the A-series matrix in the right part of the low-rank block in Figure 1

I find the size of decay_states is torch. Size([1, 24, 4, 64]).

And then according to my understanding, I actually did the following operation, the following figure shows a column vector of length 64, and then there are 4 such vectors to form a head, and then there are 24 such heads

![Image](https://github.com/user-attachments/assets/88e5c31f-f29e-4806-8cd8-2c48ef4ae8c0)

But I observed in Figure 1 that there are only two types of such column vectors actually used in Figure 1, and by analogy to 4 chunk_size, I think that there are 4 such vectors, but theoretically only the first three of them will be used

![Image](https://github.com/user-attachments/assets/046c1b7e-0d81-459e-b26a-208af8e33c46)

## 3, in mark 5: I found that this line of code is finding decay_chunk, which means to take out the largest number corresponding to the last dimension in A_cumsum [1,24,4,64], such as (a63+a62+......+a0)

i print its size ,finding decay_chunk_size = torch. Size([1, 24, 5, 5]).
 Why the dimension is 5x5 here (4 chunk_size inputs) .Shouldn't it be a 4x4 matrix, and the print found that it is such a matrix

![Image](https://github.com/user-attachments/assets/e4bcb31a-5ad4-4e55-a314-b253767abea5)

 Corresponds to Figure 1,the 5x5 matrix, and the first (0,0) value should not be 1,it should be 0.
 Then the (1,0) value should be 1 (because A2:2=1 in Figure 1), then the real value will appear in the third row, and so on.
